src/main/java/com/google/sps/utils/data/DataCollection.java

```diff
@@ -0,0 +1,235 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+
+import java.io.IOException;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+import org.jsoup.select.Elements;
+
+public class DataCollection {
+  private static final String API_URL = "https://api.gbif.org/v1/species/match?name=";
+  private static final String LIST_URL = "https://en.wikipedia.org/wiki/Lists_of_organisms_by_population";
+  private static final String LIST_CONTENT_CLASS = "mw-parser-output";
+  private static Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
+  private static KeyFactory keyFactory = datastore.newKeyFactory().setKind("Species");
+
+  public static void collectData() throws IOException {
+    List<String> urls = parseListofPages();
+    for (String url: urls) {
+      parseSpeciesTable(url);
+    }
+  }
+
+  /**
+   * Scrapes LIST_URL to get a list of all the URLs with information about species
+   * Page Structure:
+   * <ul><li>
+   *   <a>Subheading</a>
+   *   <ul>
+   *     <li><a href="LINK"></a></li>
+   *     <li><a href="LINK"></a></li>
+   *   </ul>
+   * </li></ul>
+   *
+   * @return List of all the URLs to go to
+   */
+  private static List<String> parseListofPages() throws IOException {
+    Document doc = Jsoup.connect(LIST_URL).get();
+    Elements content = doc.getElementsByClass(LIST_CONTENT_CLASS);
+    Elements listItems = content.select("li > ul > li"); // get list items that are in a nested ul
+    List<String> urls = new ArrayList<String>();
+
+    for (Element listItem: listItems) {
+
+      // Ignores list items from the Table of Contents
+      if (!listItem.attr("class").equals("")) {
+        continue;
+      }
+
+      // Gets the absolute link to the new wiki page
+      // Example: https://en.wikipedia.org/wiki/List_of_even-toed_ungulates_by_population
+      Element link = listItem.select("a[href]").first();
+      String absHref = link.attr("abs:href");
+      if (absHref.contains("#")) { // if link is actually a bookmark
+        continue;
+      }
+    
+      urls.add(absHref);
+    }
+    
+    return urls;
+  }
+
+  /**
+   * Add information for every species per url into Datastore
+   * @param url: url to parse the tables for
+   */
+  private static void parseSpeciesTable(String url) throws IOException {
+    Document doc = Jsoup.connect(url).get();
+
+    Elements table = doc.getElementsByClass("wikitable");
+    for (Element head : table.select("tbody")) {
+      for (Element row : head.select("tr")) {
+        Elements tds = row.select("td");
+
+        Species species = processSpecies(tds, url);
+        if (species == null || !species.hasNecessaryInfo()) {
+          continue;
+        }
+
+        if (!speciesAlreadyStored(species.getBinomialName())) {
+        //   System.out.println("Processing '" + species.getBinomialName() + "'...");
+          addApiInfo(species);
```

Comment:
For now, we decided that even if info about a species isn't contained in the API (aka if addApiInfo is not successful), we will still go along with adding the species to Datastore since Wikipedia provides all the necessary fields (name, image link, population, etc). Would this still apply?